internal/querycoordv2/meta/resource_manager.go

```diff
@@ -0,0 +1,562 @@
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meta
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	. "github.com/milvus-io/milvus/internal/util/typeutil"
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+)
+
+var (
+	ErrNodeAlreadyAssign            = errors.New("node already assign to other resource group")
+	ErrRGIsFull                     = errors.New("resource group is full")
+	ErrRGIsEmpty                    = errors.New("resource group is empty")
+	ErrRGNotExist                   = errors.New("resource group doesn't exist")
+	ErrRGAlreadyExist               = errors.New("resource group already exist")
+	ErrRGAssignNodeFailed           = errors.New("failed to assign node to resource group")
+	ErrRGUnAssignNodeFailed         = errors.New("failed to unassign node from resource group")
+	ErrSaveResourceGroupToStore     = errors.New("failed to save resource group to store")
+	ErrRemoveResourceGroupFromStore = errors.New("failed to remove resource group from store")
+	ErrRecoverResourceGroupToStore  = errors.New("failed to recover resource group to store")
+	ErrNodeNotAssignToRG            = errors.New("node hasn't been assign to any resource group")
+	ErrRGNameIsEmpty                = errors.New("resource group name couldn't be empty")
+	ErrDeleteDefaultRG              = errors.New("delete default rg is not permitted")
+	ErrDeleteNonEmptyRG             = errors.New("delete non-empty rg is not permitted")
+	ErrNodeNotExist                 = errors.New("node does not exist")
+	ErrNodeStopped                  = errors.New("node has been stopped")
+)
+
+var DefaultResourceGroupName = "__default_resource_group"
+
+type ResourceGroup struct {
+	nodes    UniqueSet
+	capacity int
+}
+
+func NewResourceGroup(capacity int) *ResourceGroup {
+	rg := &ResourceGroup{
+		nodes:    typeutil.NewUniqueSet(),
+		capacity: capacity,
+	}
+
+	return rg
+}
+
+// assign node to resource group
+func (rg *ResourceGroup) assignNode(id int64) error {
+	if rg.containsNode(id) {
+		return ErrNodeAlreadyAssign
+	}
+
+	rg.nodes.Insert(id)
+	rg.capacity++
+
+	return nil
+}
+
+// unassign node from resource group
+func (rg *ResourceGroup) unassignNode(id int64) error {
```

Comment:
Do these functions need to be threadsafe?

Reply:
here is a thread-safe version called `UnassignNode `